### PR TITLE
fix(server,web): rest the GitHub App install bind on the kickoff session, not the callback OAuth identity

### PR DESCRIPTION
**Pull request by Gandy2025**
## Incident

Prod onboarding: a user completed the GitHub App install (installation `139563599`, GitHub org `ShibaOrg`), but the OAuth callback refused with a raw-JSON 403 **"Not an admin of the organization this installation targets"** and the installation row sat unbound (`hub_organization_id = NULL`, confirmed in prod).

## Root cause

The callback re-resolved the acting user **from the OAuth code** (`findOrCreateUserFromGithub`) and ran the target-org admin re-check against that identity. But the browser's github.com session is independent of the First Tree session — a second GitHub account, a deleted-and-recreated account (numeric ID changes), or another person's First Tree session in the same browser all make the two identities legitimately diverge. The kickoff (`GET /orgs/:orgId/github-app-installation/install-url`) is admin-gated, so the kickoff user's authority was already proven — the callback just wasn't using it.

## Fix (4 defects)

1. **Bind authority = kickoff session.** `/install-url` signs `kickoffUserId` into the state alongside `targetOrganizationId`; the callback re-checks **that** admin live against `members` and binds on their authority. On identity mismatch it completes the bind and redirects to `next` **without issuing tokens** — signing the browser in as the foreign identity would replace the kickoff admin's session in every tab. GitHub-side authority is unchanged: `installationId` is only non-null when the OAuth user proved they administer the installed account (#423), and the nonce cookie pins the callback to the kickoff browser. Legacy states without `kickoffUserId` (≤10 min old at deploy time) keep the old behavior.
2. **No more raw-JSON dead ends.** Every browser-facing `/callback` error 302s to `/auth/github/complete#error=<code>` where the SPA renders friendly copy + a way back (codes: `state-expired`, `github-exchange-failed`, `install-not-admin`, `invite-invalid`, `invite-not-allowed`, `invite-required`, `membership-unresolved`). The most common real-world trigger — spending >10 min on GitHub's repo picker and expiring the 10-minute state JWT — now lands on a readable page. `dev-callback` keeps JSON errors (curl-able, asserted by dev tests).
3. **Observability.** The refusal path logs structured warns with both identities (`github_app.install_callback_admin_check_failed` / `github_app.install_callback_identity_mismatch`) — this incident was undiagnosable from logs (the 403 path logged nothing).
4. **Copy.** The refusal names the First Tree team; "organization" read as the GitHub org to everyone who saw it.

## Tests (red → green)

- `binds via the kickoff session's authority when the callback GitHub identity differs` — the incident class; failed 403 before, passes with the row bound and no token fragment in the redirect.
- `redirects to a friendly error page when the kickoff admin's membership was revoked mid-flight` — the legitimate refusal keeps refusing, now on the SPA surface.
- Updated: legacy-state refusals + state-rejection tests assert the redirect contract; `install-url` test pins `kickoffUserId` + `targetOrganizationId` in the signed state; web test covers the `#error=` fragment rendering.
- Full suite green: server 1352, web 743, cli 436, client 1197, shared 408. `pnpm check` + `pnpm typecheck` clean.

## Recovery for the affected user

After deploy, re-running "Install on GitHub" from Settings/onboarding completes and binds installation `139563599` (GitHub redirects straight back with the existing install). No manual DB surgery needed.

Tree PR (merged): agent-team-foundation/first-tree-context#492

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/server/src/api/auth/github.ts [hunk 1]
```diff
@@ -386,14 +447,65 @@ async function completeOauthFlow(
...
+          targetOrganizationId,
+          kickoffUserId: bindAuthorityUserId,
+          oauthUserId: userId,
+          githubId: profile.githubId,
+          githubLogin: profile.login,
+          installationId,
+        },
+        "install callback: OAuth identity differs from the kickoff session — binding on kickoff authority, skipping sign-in",
+      );
+      if (installationId !== null) {
```
Comment (yuezengwu) [inline packages/server/src/api/auth/github.ts]: This branch needs to fail before redirecting to `next` when `installationId` is `null`. In this target-org flow, `installationId` can be missing from the callback or cleared above when the GitHub admin proof / metadata upsert fails. With a mismatched OAuth identity, we then skip token issuance and still `302` to success here; for onboarding's popup that means `/onboarding/connected` renders "Connected" and auto-closes while no `github_app_installations` row is bound, so the original tab just keeps polling/waiting. Please redirect to the SPA error surface instead (and add a regression test for `targetOrganizationId + kickoffUserId + foreign OAuth identity + null/cleared installationId`).

**Review by yuezengwu — CHANGES_REQUESTED** [review]
Reviewed the PR goal and core change: the install bind now rests on the kickoff First Tree session while GitHub-side install authority still comes from the callback OAuth user, and browser-facing callback failures move to the SPA error surface. That matches the updated Context Tree direction.

I found one request-changes issue: the identity-mismatch branch can still report success without any verified installation to bind. No database/schema change in this PR, but this does touch core auth and org-binding behavior, so I would fix this before landing.

**Review by baixiaohang — APPROVED** [review]
## Review by codex-assistant - PR #988

**建议**: approve
- 理由: 绑定 authority 改为 kickoff session 后仍保持 GitHub-side admin proof、live membership re-check 和 no-token mismatch redirect，未发现阻塞回归。

**分级**: B-low
- Path 初判: 涉及 `packages/server/**`，未命中 migrations / prisma schema / chat-message-session-agent service 兜底；web 改动为 OAuth complete 页面 -> B-low。
- 语义升级: 无。

**PR 摘要**
- 作者 / repo: Gandy2025 / agent-team-foundation/first-tree
- 解决问题: GitHub App install callback 用 OAuth identity 做 First Tree admin re-check，导致 kickoff admin 与 callback GitHub identity 不一致时安装成功但 Hub 绑定失败。
- 设计方案: `/install-url` 将 `targetOrganizationId` 与 `kickoffUserId` 写入 signed state；callback 继续用 OAuth user 证明 GitHub installation admin 权限，但 First Tree org bind authority 改为 re-check kickoff user 的 live admin membership；identity mismatch 时完成 bind 并直接回 `next`，不下发 session tokens；browser callback 错误统一重定向到 SPA error surface。
- 受影响模块: `packages/server/src/api/auth/github.ts`, `packages/server/src/api/orgs/github-app.ts`, `packages/server/src/services/oauth-state.ts`, `packages/web/src/pages/oauth-complete.tsx`。

**Review 意见**
✅ 1. authority 分层是清晰的：GitHub installation admin proof 仍绑定 OAuth user + `installation_id`，First Tree org bind proof 改为 kickoff user 的 live membership，符合 tree 中 GitHub App install-bind 约束。
✅ 2. identity mismatch 分支在 bind 后不签发 First Tree tokens，避免把 kickoff admin 的浏览器会话替换成 callback OAuth identity。
✅ 3. legacy state、revoked admin、state rejection 和 web `#error=` 渲染都有对应覆盖；我按 review 规则未运行本地检查或单测。

**已执行**
- 已 approve。

**Review by baixiaohang — CHANGES_REQUESTED** [review]
## Review by codex-assistant - PR #988

**建议**: request_changes
- 理由: identity-mismatch 分支在没有 verified installation 可绑定、或绑定失败时仍会按成功跳回 `next`，与“完成 bind 后不下发 tokens”的安全契约不一致。

**分级**: B-low
- Path 初判: 涉及 `packages/server/**`，未命中 migrations / prisma schema / chat-message-session-agent service 兜底；web 改动为 OAuth complete 页面 -> B-low。
- 语义升级: 无。

**PR 摘要**
- 作者 / repo: Gandy2025 / agent-team-foundation/first-tree
- 解决问题: GitHub App install callback 用 OAuth identity 做 First Tree admin re-check，导致 kickoff admin 与 callback GitHub identity 不一致时安装成功但 Hub 绑定失败。
- 设计方案: `/install-url` 将 `targetOrganizationId` 与 `kickoffUserId` 写入 signed state；callback 继续用 OAuth user 证明 GitHub installation admin 权限，但 First Tree org bind authority 改为 re-check kickoff user 的 live admin membership；identity mismatch 时不下发 session tokens；browser callback 错误统一重定向到 SPA error surface。
- 受影响模块: `packages/server/src/api/auth/github.ts`, `packages/server/src/api/orgs/github-app.ts`, `packages/server/src/services/oauth-state.ts`, `packages/web/src/pages/oauth-complete.tsx`。

**Review 意见**
❌ 1. identity-mismatch 分支把 bind 变成了可选 side effect，但仍返回成功 redirect。 [R4 / `packages/server/src/api/auth/github.ts:498`]

`installationId` 在进入 `completeOauthFlow` 前可能已经是 `null`：callback query 里没有 `installation_id`，或 GitHub-side admin proof / fetch / upsert 失败后被 fail-closed 清空。这个分支现在只在 `installationId !== null` 时尝试 `bindInstallationToOrg`，并且即使 bind 抛错也只是 warn，随后无条件 `return reply.redirect(next, 302)`。

这会让 kickoff admin 在 identity mismatch 场景里看到一次“成功”的 `/onboarding/connected` 或 Settings 回跳，但实际没有任何 installation 被绑定；同时也没有 token fragment，所以 SPA 只是沿用旧会话，错误被隐藏。该分支的 no-token success 应该只发生在 verified installation row 已经实际绑定成功之后；否则应走 friendly error surface（或至少不要跳到 success `next`）。

✅ 2. 其它核心分层是合理的：GitHub installation admin proof 仍绑定 OAuth user + `installation_id`，First Tree org bind proof 改为 kickoff user 的 live membership，且 mismatch 成功路径不签发 First Tree tokens。

**已执行**
- 已 request_changes；这条 review 取代我刚才过早提交的 approve。

**Comment by Gandy2025** [comment]
Addressed the request-changes finding (independently confirmed by three reviewers: @yuezengwu, codex-assistant, and a codex CLI review — all converged on the same line).

**Fix (af52b78e):** the identity-mismatch branch's no-token success redirect is now earned only by an actual bind:
- no verified installation (`installation_id` absent, or cleared by the fail-closed GitHub-side admin proof) → `#error=install-not-verified`
- `bindInstallationToOrg` refusal (e.g. D2 1:1 already-bound-elsewhere) → `#error=install-bind-failed`

Both legs covered by red→green tests (`surfaces an error (not success) when the identities mismatch and …`); friendly copy added for the two new codes. Server target suite 8/8, web DOM suite 20/20, `pnpm check` + `typecheck` clean.

**Review by yuezengwu — APPROVED** [review]
Re-reviewed the synchronized head `af52b78e`.

The previous request-changes issue is fixed: the identity-mismatch path now earns the no-token success redirect only after a real `bindInstallationToOrg` succeeds. Missing/cleared `installationId` now redirects to `install-not-verified`, bind failures redirect to `install-bind-failed`, and both legs have regression coverage.

No further blocking findings from this pass. Local tests were not run because this temporary worktree has no `node_modules`; GitHub CI is in progress and I did not wait on it per review policy.

**Review by baixiaohang — CHANGES_REQUESTED** [review]
## Review by codex-assistant - PR #988 follow-up

**建议**: request_changes
- 理由: 上一轮 “无实际 bind 仍 success redirect” 的核心问题已修，但新的错误页仍会把 onboarding popup 的 `next=/onboarding/connected` 暴露成返回链接，用户点击后会进入成功/自动关闭页。

**分级**: B-low
- Path 初判: 涉及 `packages/server/**`，未命中 migrations / prisma schema / chat-message-session-agent service 兜底；web 改动为 OAuth complete 页面 -> B-low。
- 语义升级: 无。

**PR 摘要**
- 作者 / repo: Gandy2025 / agent-team-foundation/first-tree
- 解决问题: GitHub App install callback 的 First Tree bind authority 改为 kickoff session，并给 browser-facing callback 失败提供 SPA 错误页。
- 设计方案: 最新 head `af52b78` 让 identity-mismatch 分支只有在 verified installation row 实际 bind 成功后才 no-token redirect；无 verified install 与 bind failure 分别返回 `install-not-verified` / `install-bind-failed`。
- 受影响模块: `packages/server/src/api/auth/github.ts`, `packages/web/src/pages/oauth-complete.tsx`, `packages/server/src/__tests__/github-app-callback-target-org.test.ts`。

**Review 意见**
❌ 1. install error fragment still carries the popup success page as its “back” target. [R4 / `packages/server/src/api/auth/github.ts:509`]

The new no-verified-install test intentionally uses `next=/onboarding/connected`, which is the popup landing page whose UI says “Connected” and auto-closes the tab. The server now redirects to `/auth/github/complete#error=install-not-verified&next=/onboarding/connected`, and `OAuthCompletePage` turns that `next` into the visible `Back to First Tree` link. If the user clicks it after seeing the error, they are sent to the same success/auto-close page that this fix is trying to avoid.

For install failure codes, the error escape hatch should not preserve `/onboarding/connected`; normalize it to a real recovery surface such as `/onboarding`, `/settings/github`, or omit `next` so the default `/` is used. Please add coverage that the `install-not-verified` / bind-failed error page link does not point at `/onboarding/connected`.

✅ 2. The actual bind gate is now correct: no-token success is reached only after `bindInstallationToOrg` succeeds, while missing/cleared `installationId` and bind refusal both land on the SPA error surface.

**已执行**
- 已 request_changes 最新 head `af52b78`；按 PR review 规则未运行本地检查或单测。

**Review by yuezengwu — CHANGES_REQUESTED** [review]
Changing my previous approval to request changes after the follow-up review.

The bind gate itself is fixed, but the new install failure errors still preserve `next=/onboarding/connected` for the onboarding popup path. `OAuthCompletePage` renders that as the visible "Back to First Tree" link, so a user who sees `install-not-verified` / `install-bind-failed` can click straight into the success/auto-close sentinel page. That keeps a false-success recovery path in the exact install failure surface this PR is trying to make understandable.

Please normalize or omit `next` for install failure codes when it points at `/onboarding/connected` (for example, fall back to `/onboarding`, `/settings/github`, or `/`), and add coverage that these error links do not target `/onboarding/connected`.

**Comment by Gandy2025** [comment]
Round-2 finding addressed in 702a2de7 (@yuezengwu, codex-assistant — converged again, thanks):

- `redirectCallbackError` now normalizes `next=/onboarding/connected` → `/onboarding` for **every** error code, so the error page's visible "Back to First Tree" link can never land on the auto-close "Connected" sentinel. Settings-flow `next` values pass through unchanged.
- Coverage added as requested: `install-not-verified` pins `next=/onboarding` (red→green), `install-bind-failed` pins the `/settings/github` passthrough.

Also: the CodeQL high (`js/missing-rate-limiting` on `/callback`) is a false positive — `@fastify/rate-limit` is registered app-wide (`app.ts:371`, max 3000/window) and per-route config was deliberately removed in #984 (merged today). Dismissed with that rationale.

**Review by yuezengwu — APPROVED** [review]
Re-reviewed synchronized head `702a2de7`.

The follow-up request-changes issue is fixed. `redirectCallbackError` now normalizes `/onboarding/connected` to `/onboarding` before putting `next` into the error fragment, so the error page's visible "Back to First Tree" link no longer targets the popup success/auto-close sentinel. The new server assertions cover both sides: the onboarding popup error path normalizes to `/onboarding`, while the Settings recovery path preserves `/settings/github`.

No further blocking findings. I ran `git diff --check` locally; targeted tests were not run because this temporary worktree has no `node_modules`. GitHub CI is in progress and I did not wait on it per review policy.

**Review by baixiaohang — APPROVED** [review]
## Review by codex-assistant - PR #988 follow-up 2

**建议**: approve
- 理由: 最新 head `702a2de` 已解决上一轮 false-success recovery path；install failure 的错误页返回链接不再指向 `/onboarding/connected`，核心 bind gate 也保持只在实际绑定成功后才 no-token success。

**分级**: B-low
- Path 初判: 涉及 `packages/server/**`，未命中 migrations / prisma schema / chat-message-session-agent service 兜底；web 改动为 OAuth complete 页面 -> B-low。
- 语义升级: 无。

**PR 摘要**
- 作者 / repo: Gandy2025 / agent-team-foundation/first-tree
- 解决问题: GitHub App install callback 的 First Tree bind authority 改为 kickoff session，并把 browser-facing callback 失败导向 SPA 错误页而不是 raw JSON 或 false-success 页面。
- 设计方案: `targetOrganizationId` + `kickoffUserId` 进入 signed state；callback 用 OAuth user 证明 GitHub installation admin 权限，用 kickoff user re-check First Tree admin membership；identity mismatch 时只有实际 `bindInstallationToOrg` 成功才直接回 `next`，否则进入 typed error；error recovery link 将 `/onboarding/connected` 规范化为 `/onboarding`。
- 受影响模块: `packages/server/src/api/auth/github.ts`, `packages/server/src/api/orgs/github-app.ts`, `packages/server/src/services/oauth-state.ts`, `packages/web/src/pages/oauth-complete.tsx`。

**Review 意见**
✅ 1. 上一轮 blocker 已解决：`redirectCallbackError` 现在把 `/onboarding/connected` 规范化为 `/onboarding`，错误页的 visible recovery link 不会再进入 Connected auto-close sentinel。 [R4 / `packages/server/src/api/auth/github.ts`]
✅ 2. 覆盖与行为匹配：`install-not-verified` 断言 `next=/onboarding`，`install-bind-failed` 断言 Settings flow 的 `/settings/github` passthrough；实际 bind gate 仍要求 `bindInstallationToOrg` 成功后才 no-token success。 [R4 / `packages/server/src/__tests__/github-app-callback-target-org.test.ts`]
⚠️ 3. 我未运行本地检查或单测；按 PR Review 模式只做静态复核。远端 CI 在我提交 review 前仍有部分 job in progress。

**已执行**
- 已 approve 最新 head `702a2de`。
